src/tiledimage.js

```diff
@@ -2018,28 +2090,95 @@ $.extend($.TiledImage.prototype, $.EventSource.prototype, /** @lends OpenSeadrag
-            _this = this;
+     * @param {*} data image data, the data sent to ImageJob.prototype.finish(), by default an Image object,
+     *   can be null: in that case, cache is assigned to a tile without further processing
+     * @param {?Number} cutoff ignored, @deprecated
+     * @param {?XMLHttpRequest} tileRequest
+     * @param {?String} [dataType=undefined] data type, derived automatically if not set
+     */
+    _setTileLoaded: function(tile, data, cutoff, tileRequest, dataType) {
+        tile.tiledImage = this; //unloaded with tile.unload(), so we need to set it back
+        // does nothing if tile.cacheKey already present
+
+        let tileCacheCreated = false;
+        tile.addCache(tile.cacheKey, () => {
+            tileCacheCreated = true;
+            return data;
+        }, dataType, false, false);
+
+        let resolver = null,
+            increment = 0,
+            eventFinished = false;
+        const _this = this,
+            now = $.now();
+
+        function completionCallback() {
+            increment--;
+            if (increment > 0) {
+                return;
+            }
+            eventFinished = true;
+
+            //do not override true if set (false is default)
+            tile.hasTransparency = tile.hasTransparency || _this.source.hasTransparency(
+                undefined, tile.getUrl(), tile.ajaxHeaders, tile.postData
+            );
+            // tile.updateRenderTarget(true);
+            // //make sure cache data is ready for drawing, if not, request the desired format
+            // const cache = tile.getCache(),
+            //     requiredTypes = _this._drawer.getSupportedDataFormats();
+            // if (!cache) {
+            //     $.console.warn("Tile %s not cached or not loaded at the end of tile-loaded event: tile will not be drawn - it has no data!", tile);
+            //     resolver();
+            // } else if (!requiredTypes.includes(cache.type)) {
+            //     //initiate conversion as soon as possible if incompatible with the drawer
+            //     //either the cache is a new item in the system (do process), or the cache inherits data from elsewhere (no-op),
+            //     // or the cache was processed in this call
+            //     tile.transforming = now;  // block any updates on the tile
+            //     cache.prepareForRendering(_this._drawer.getId(), requiredTypes, _this._drawer.options.usePrivateCache).then(cacheRef => {
+            //         if (!cacheRef) {
+            //             return cache.transformTo(requiredTypes);
+            //         }
+            //         if (tile.processing === now) {
+            //             tile.updateRenderTarget();
+            //         }
+            //         return cacheRef;
+            //     }).then(resolver);
+            // } else {
+            //     resolver();
+            // }
+
+            // TODO consider first running this event before we call tile-loaded...
```

Comment:
Here it is nicely visible how the separation clears up all the logics. One thing to consider is whether to move the invalidation event before tile loaded event. Also, people will come to work with `tile.loaded=tile.loading=false` state, so this might need some mention in docs as well as ensure no tile logics blocks functional behavior here... I left commented code here in case we might want to revert, but it seems to be a better direction...

Reply:
This does seem cleaner! 

I would think it makes sense to have the invalidation event before the tile loaded event.

Reply:
Moved